### PR TITLE
adds constructors to docs

### PR DESCRIPTION
**Pull request by stevekrouse**
I put it in the order I prefer:

* constructors
* `isThing()`
* all other functions

### CODE REVIEW

**Comment by stevekrouse** [comment]
Oops, `Stream.of` doesn't appear to be a function...

File: README.md [hunk 1]
```diff
@@ -429,17 +429,37 @@ misbehavingStream.log();
 
 ### Future
 
-#### `Future#listen(o: Consumer<A>): void`
+#### `Future.of(a: A): Future<A>`
 
-Adds a consumer as listener to a future. If the future has already
-occurred the consumer is immediately pushed to.
+Converts any value into a future that happens immediately.
```
Comment (paldepind) [inline README.md]: I'll have to be a bit nitpicky about the wording here.

Semantically `Future.of(a)` is equivalent to `(-Infinity, a)`. This means that `Future.of` returns a future that has always occurred. The wording "happens immediately" to me sounds as if `Future.of` returns a future that occurs at the time where `Future.of` is called. This however, would not be a pure function.

I hope that makes sense. I think the wording should be changed to something like

> Returns a future that has already occurred with the given value.

Or maybe

> Returns a future that has at any point in time always occurred with the given value.

 However, these may be to "technical" or unintuitive. So let me know if you have a better idea :smile:

File: README.md [hunk 2]
```diff
@@ -429,17 +429,37 @@ misbehavingStream.log();
...
+
 ### Stream
 
+#### `empty: Stream<A>`
+
+Empty stream.
+
+#### `Stream.of(a: A): Stream<A>`
+
+Converts any value into a constant stream that happens immediately.
```
Comment (paldepind) [inline README.md]: As you said, this function doesn't exist and there is no good way to implement it.

File: README.md [hunk 3]
```diff
@@ -429,17 +429,37 @@ misbehavingStream.log();
...
+Returns `true` if `f` is a future and `false` otherwise.
+
+#### `Future#listen(o: Consumer<A>): void`
+
+Adds a consumer as listener to a future. If the future has already
+occurred the consumer is immediately pushed to.
+
 ### Stream
 
+#### `empty: Stream<A>`
```
Comment (paldepind) [inline README.md]: If I recall correctly it has the type `Stream<any>`.

File: README.md [hunk 4]
```diff
@@ -566,6 +586,20 @@ Returns a stream that after occurring, ignores the next occurrences in
 
 ### Behavior
 
+#### `Behavior.of(a: A): Behavior<A>`
```
Comment (paldepind) [inline README.md]: The generic is missing. 

```diff
+  #### `Behavior.of<A>(a: A): Behavior<A>`
- #### `Behavior.of(a: A): Behavior<A>`
```

File: README.md [hunk 5]
```diff
@@ -631,6 +655,10 @@ The Now monad represents a computation that takes place in a given
 moment and where the moment will always be now when the computation is
 run.
 
+#### `Now.of(a: A): Now<A>`
```
Comment (paldepind) [inline README.md]: The generic is missing.

**Comment by paldepind** [comment]
Thanks a lot for the PR :+1:

Overall this looks good but I've left a few comments regarding some smaller details :smile:

File: README.md (hunk 1 above)
Comment (stevekrouse) [reply README.md]: Ah, makes sense. Thanks! I much prefer to over-explain than under-explain a concept, so I'll try to incorporate a lot of what you said above into the docs...

File: README.md (hunk 3 above)
Comment (stevekrouse) [reply README.md]: Ah, sorry, I'm new to TypeScript (from Haskell) and to me those seem like the same type

File: README.md (hunk 2 above)
Comment (stevekrouse) [reply README.md]: That's un-intuitive to me. I'd love to put in the documentation why this doesn't exist. For example, why can't it have the same semantics at `Future.of` of `(-Infinity, a)`?

File: README.md (hunk 4 above)
Comment (stevekrouse) [reply README.md]: ok, thanks!

File: README.md (hunk 5 above)
Comment (stevekrouse) [reply README.md]: thanks!

File: README.md (hunk 2 above)
Comment (paldepind) [reply README.md]: That's a good question! You're right that we could impleent `Stream.of(a)` to be equivalent to `[(-Infinity, a)]`. The `Stream.of` function could actually work like that.

But, here's the catch: Hareactive never makes it possible to access a streams past occurrences. Said another way, it is impossible to write code that when executed depends on occurrences that happened earlier. There are two reasons for this.

* It makes it possible to forget occurrences as soon as they occur. We just need to notify any listeners and then the data in the occurrence can be garbage collected.
* It's not very useful. In the cases where one would want to know the last occurrence using a behavior is usually the better choice.

In the light of this the `[(-Infinity, a)]` implementation is not useful. It returns a stream of an occurrence that is always in the past and hence an occurrence that is inaccessible. We could actually implement `of` on `Stream` like this:

```
of<A>(a: A): Stream<A> {
  return empty;
}
```

That would be a valid implementation because the occurrence is in the past, and hence inaccessible anyway.

To sum it up, we could implement `Stream.of` but it would not be useful.

Note that `Future` is different. Here we keep the data once the future has occurred.

File: README.md (hunk 3 above)
Comment (paldepind) [reply README.md]: I see what you mean :smile:. I actually remember being slightly confused when I wrote `empty`. I tried to use a generic. But TypeScript only allows generics on functions and it turned out that `any` works best anyway.

File: README.md (hunk 2 above)
Comment (stevekrouse) [reply README.md]: Cool, makes sense! Appreciate this. 

Why do we keep data once `Future` has occurred?

Comment (stevekrouse) [reply README.md]: Nevermind, I can see why that would make sense for `Future`s

**Comment by paldepind** [comment]
Looks great! Thanks a lot.
